Fix concurrency issue in CastleDynamicProxyGenerator

thomaslevesque opened the pull request:
Fixes #773

blairconrad commented:
I don't mean to quibble, but we don't _know_ that this fixes the issue that @apacha sees. I believe it might, and it's a good step, but maybe we should instead claim that we're no longer using a static ProxyGenerationOptions?

blairconrad commented:
Also, bigger change than I anticipated. I won't get to for several hours. Sorry.

thomaslevesque commented:
> maybe we should instead claim that we're no longer using a static ProxyGenerationOptions?

True. I'll change the description.

> Also, bigger change than I anticipated. I won't get to for several hours. Sorry.

No worry, it can wait until then.

It's not as big as it seems; I'll put a summary in the PR description.

blairconrad commented:
>  It's not as big as it seems; I'll put a summary in the PR description.

Oh, I'll never turn down a summary, but I'm not super scared by it. It's just that there's the "2 minute review while compiling" and then there's the "wait until you're not at work review".

blairconrad commented on src/FakeItEasy/Creation/CastleDynamicProxy/CastleDynamicProxyGenerator.cs:
```diff
@@ -67,14 +89,11 @@ public ProxyGeneratorResult GenerateProxy(Type typeOfProxy, IEnumerable<Type> ad
 
+            Guard.AgainstNull(additionalInterfacesToImplement, "additionalInterfacesToImplement");
             GuardAgainstConstructorArgumentsForInterfaceType(typeOfProxy, argumentsForConstructor);
+            Guard.AgainstNull(fakeCallProcessorProvider, "fakeCallProcessorProvider");
```
The order of the Guard/GuardAgainstConstructorArgumentsForInterfaceType/Guard is a little peculiar.
Also, since we're not dereferencing additionalInterfacesToImplement or fakeCallProcessorProvider, I'd've expected to see the Guards deeper in. 
Functionally, none of this makes a difference

thomaslevesque replied:
It's the same order as the parameters, but I can change it. Anyway, seeing it again, I find it a bit strange too... I'll make another pass at it, and try to move the guard clauses to the method where the parameters are dereferenced.

adamralph replied:
Having guard clauses for arguments half way through methods is also a massive code smell.

blairconrad replied:
It looked weird to me too, but I felt better when I noticed that we were short-circuiting above. I guess pushing the guards into `CreateProxyGeneratorResult` will help with this as well.

thomaslevesque replied:
> Having guard clauses for arguments half way through methods is also a massive code smell.

The `GuardAgainstConstructorArgumentsForInterfaceType` was already in the middle, it's why I put the others there. I think it was an optimization to bail early if the type is a value type or is sealed. But it's not a very useful optimization, so I'll just put everything at the start of the method.

adamralph commented:
> maybe we should instead claim that we're no longer using a static ProxyGenerationOptions?

:+1: Given the non-deterministic nature of concurrency bugs, perhaps we should re-title this "Improve thread safety by removing shared ProxyGenerationOptions" and label it as an independent `enhancement`. After release, and given some time, we can see if #773 is still an issue, and if not, close it as `invalid`, i.e. "not a bug". It's weird, but then so is the non-deterministic nature of concurrency bugs.

The alternative is to say that this PR "fixes #773" if we are reasonably confident it does, and then, after some time, if there are still problems, we create a new bug issue to be fixed and attempt it to fix it again.

blairconrad commented:
> The alternative is to say that this PR "fixes #773" if we are reasonably confident it does, and then, after some time, if there are still problems, we create a new bug issue to be fixed and attempt it to fix it again.

I did not think that this would be my preference. Only once I read it, it is. I think @thomaslevesque nailed the analysis, and this is the cause of the problem. @apacha's [comment](https://github.com/FakeItEasy/FakeItEasy/issues/773#issuecomment-227170076) that some of his tests spawn background tasks only increases my confidence.

Sorry to flip-flop on this, but I do believe that this will resolve the problem.

adamralph commented:
OK, I'm also convinced.

I've taken the liberty to change the link to #773 in the description to "Fixes #773" and moved #773 into progress.

thomaslevesque commented:
I restored the guard clauses to their previous state (except a redundant one that I removed). Not sure what I had in mind when I changed them...
The only other change is that I inlined `GetConstructorArgumentsArray`, using the null propagation operator.

adamralph commented on src/FakeItEasy/Creation/CastleDynamicProxy/CastleDynamicProxyGenerator.cs:
```diff
@@ -40,18 +40,42 @@ public CastleDynamicProxyGenerator(CastleDynamicProxyInterceptionValidator inter
+        private static ProxyGenerationOptions CreateProxyGenerationOptions()
+        {
+            return new ProxyGenerationOptions(ProxyGenerationHook);
+        }
```
Can we inline this?

thomaslevesque replied:
We could, but why? I created a method because it was used in two places; if, for some reason or another, we need to make changes to the `ProxyGenerationOptions`, there will be only one place to change.

adamralph replied:
YAGNI though. Do we really need a private method around a call to a constructor?

adamralph replied:
Anyway, it's a minor point. I won't insist on it.

thomaslevesque replied:
Well, unless it really bothers you, I'd rather keep it. Looks cleaner to me. And it will be inlined at runtime anyway ;)

adamralph commented:
Thanks @thomaslevesque!

thomaslevesque commented:
Thanks for the merge!
